pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh

```diff
@@ -68,6 +69,14 @@ while [ "$#" -gt 0 ]; do
       --rollback)
         rollback=1
         ;;
+      --generation)
+        if [ $# -eq 0 ]; then
+            log "Must provide a generation number"
+            log "Usage: \`$(basename "$0") ${action:-switch} --generation <NUMBER>'"
```

Comment:
Why do I need to give an action when I just want to switch to a generation?

Reply:
We don't know if you want to switch, boot or test said generation, so you have to tell us what you want to do with it...

Reply:
but if I want to switch to an older generation, why does it start to build the current config file from disk which might be newer than the current generation?

Reply:
That is interesting - I don't think this happened to me before :thinking: Where did you see this behavior? If it is in the tests: the first three tests don't actually do the `--generation`, but they create the generations, so there it is expected that they build the config files...

I know it isn't a good answer, but: I don't really know what I am doing (at least not 3 years after I first wrote this patch). I just adapted the code for the `--rollback` since I thought they would be more or less identical... I have forgotten the few things I knew about the nix-env stuff and the rest on how this is activated :see_no_evil: If we just wanted to activate the generation and don't care about the action I guess we could also just run `/nix/var/nix/profiles/system-$generation-link/activate`...

I tried to add a test for the `boot --generation`, but something is strange (maybe the QEMU runs some kind of tempfs - at least after the `machine.reboot()` and it doesn't have nix-channels or actually the old generations any more on the disk) and so I could only check that the current generation was still active...

Reply:
I could not reproduce the building of the current configuration.nix in the tests, as this passes, which I think is what you mean:
```diff
diff --git a/nixos/tests/nixos-rebuild-switch.nix b/nixos/tests/nixos-rebuild-switch.nix
index 8fbd1f2eb587..c726dcca7b83 100644
--- a/nixos/tests/nixos-rebuild-switch.nix
+++ b/nixos/tests/nixos-rebuild-switch.nix
@@ -89,7 +89,9 @@ import ./make-test-python.nix ({ pkgs, ... }: {
           ${testForActiveGeneration "3"}

       with subtest("must roll back with `switch --generation`"):
-          machine.succeed("nixos-rebuild switch --generation 1")
+          ${createConfig "4"}
+          out = machine.succeed("nixos-rebuild switch --generation 1")
+          assert "building the system configuration..." not in out
           ${testForBootGeneration "1"}
           ${testForActiveGeneration "1"}
```
Did I misunderstand you or do you have a reproduction by any chance?

Reply:
well the script (re)executes itself https://github.com/NixOS/nixpkgs/blob/1860f2ab371659d417f30e9f32e242c6ce5928a5/pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh#L405C68-L405C75, so maybe that is why it starts building?

Reply:
> If we just wanted to activate the generation and don't care about the action I guess we could also just run `/nix/var/nix/profiles/system-$generation-link/activate`...

Yes, this is basically what I wanted to do.

> Did I misunderstand you or do you have a reproduction by any chance?

I think this is what I did, yes.

> well the script (re)executes itself

I am usually passing --fast to it.

Reply:
> > If we just wanted to activate the generation and don't care about the action I guess we could also just run `/nix/var/nix/profiles/system-$generation-link/activate`...
> 
> Yes, this is basically what I wanted to do.

I just implemented it with "feature parity" to the `--rollback`. I personally mostly use `switch` as an action, but maybe e.g. someone could want to set a "known good generation" as a default after a reboot when testing a server update and on error they can just do a reboot and everything should be back to normal. Could something like this be a use case (and if not, just out of curiosity: why would it be one for rollback)?

> > Did I misunderstand you or do you have a reproduction by any chance?
> 
> I think this is what I did, yes.

Very interesting... This should not happen (https://github.com/NixOS/nixpkgs/pull/105910/files#diff-fc23c722779993fcffc932965cd143b265aa196d577ca10173b63e5bdcbc0001R608-R609) and also the test that I wrote above passes... 🤔 I really don't have a clue how this could have happened... Anyone has any idea (or even better can add a failing test case)?